clients/src/main/java/org/apache/kafka/common/protocol/ObjectSizeCache.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.protocol;
+
+import java.util.IdentityHashMap;
+
+/**
+ * The ObjectSizeCache stores the sizes of objects.
+ * It is intended to be used as part of a two-pass serialization process like:
+ * ObjectSizeCache cache = new ObjectSizeCache();
+ * message.calculateSize(version, cache);
+ * message.write(version, cache
+ */
+public final class ObjectSizeCache {
```

Comment:
It looks like this map also stores byte arrays, so we should probably change the name and documentation?

Reply:
arrays are object, right? :)
What names would be better?  SerializationSizeCache?  Just SizeCache?

Reply:
No, I mean that we are storing the actual array as the value, not a size:

```java
public void cacheSerializedValue(Object o, byte[] val) {
    map.put(o, val);
}
```

Reply:
Good point.  I renamed this class to "ObjectSerializationCache" which should be clearer.